traitsui/tests/test_tools.py

```diff
@@ -0,0 +1,131 @@
+    is_current_backend_wx,
+    skip_if_not_qt4,
+    skip_if_not_wx,
+    process_cascade_events,
+)
+
+
+if is_current_backend_qt4():
+
+    # Create a QObject that will emit a new event to itself as long as
+    # it has not received enough.
+
+    from pyface.qt import QtCore
+
+    class DummyQObject(QtCore.QObject):
+
+        def __init__(self, max_n_events):
+            super().__init__()
+            self.max_n_events = max_n_events
+            self.n_events = 0
+
+        def event(self, event):
+            self.n_events += 1
+
+            if self.n_events < self.max_n_events:
+                new_event = QtCore.QEvent(QtCore.QEvent.User)
+                QtCore.QCoreApplication.postEvent(self, new_event)
+            return True
+
+
+if is_current_backend_wx():
+
+    # Create a wx.EvtHandler that will emit a new event to itself as long as
+    # it has not received enough.
+
+    import wx
+    import wx.lib.newevent
+
+    NewEvent, EVT_SOME_NEW_EVENT = wx.lib.newevent.NewEvent()
+
+    class DummyWxHandler(wx.EvtHandler):
+
+        def __init__(self, max_n_events):
+            super().__init__()
+            self.max_n_events = max_n_events
+            self.n_events = 0
+
+        def TryBefore(self, event):
+            self.n_events += 1
+            if self.n_events < self.max_n_events:
+                self.post_event()
+            return True
+
+        def post_event(self):
+            event = NewEvent()
+            wx.PostEvent(self, event)
+
+
+class TestProcessEventsRepeated(unittest.TestCase):
+    """ Test process_events actually processes all events, including the events
+    posted by the processed events.
+    """
+
+    @skip_if_not_qt4
+    def test_qt_process_events_process_all(self):
+        from pyface.qt import QtCore
+
+        def cleanup(q_object):
+            q_object.deleteLater()
+            # If the test fails, run process events at least the same
+            # number of times as max_n_events
+            for _ in range(q_object.max_n_events):
+                QtCore.QCoreApplication.processEvents(
+                    QtCore.QEventLoop.AllEvents
+                )
+
+        max_n_events = 10
+        q_object = DummyQObject(max_n_events=max_n_events)
+        self.addCleanup(cleanup, q_object)
+
+        QtCore.QCoreApplication.postEvent(
+            q_object, QtCore.QEvent(QtCore.QEvent.User)
+        )
+
+        # sanity check calling processEvents does not process
+        # cascade of events.
+        QtCore.QCoreApplication.processEvents(QtCore.QEventLoop.AllEvents)
+        self.assertEqual(q_object.n_events, 1)
+
+        # when
+        process_cascade_events()
+
+        # then
+        self.assertEqual(q_object.n_events, max_n_events)
+
+    @skip_if_not_wx
+    def test_wx_process_events_process_all(self):
+
+        def cleanup(wx_handler):
+            GUI.process_events()
+            wx_handler.Destroy()
+
+        wx_handler = DummyWxHandler(max_n_events=10)
```

Comment:
Nit: could we have a local variable `max_n_events` as in `test_qt_process_events_process_all`?